x-pack/docs/en/security/es-security-principles.asciidoc

```diff
@@ -0,0 +1,55 @@
+[[es-security-principles]]
+== {es} security principles
+Protecting your {es} cluster and the data it contains is of utmost importance.
+There are numerous security practices and considerations for keeping your
+cluster safe, such as not saving passwords in plain text files or making sure
+that users have appropriate permissions (most users shouldn't be a superuser).
+These guidelines are important, but the following information provides something
+greater. These principles are foundational to running {es} in a secure manner.
+
+[discrete]
+[[security-dont-run-without]]
+=== Run {es} with security enabled
+Never run an {es} cluster without security enabled. This principle cannot be
+overstated. Running {es} without security leaves your cluster exposed to anyone
+on the internet, permitting anyone to download, modify, or delete any data in
+your cluster.
+<<configuring-stack-security,Start the {stack} with security enabled>> or
+<<manually-configure-security,manually configure security>> to prevent
+unauthorized access to your clusters and ensure that internode communication
+is secure. 
+
+[discrete]
+[[security-dont-expose]]
+=== Protect {es} from public internet traffic
+Even with security enabled, never expose {es} to public internet traffic.
+Using an application to sanitize requests to {es} still poses risks, such as
+a malicious user writing <<search,`_search`>> requests that could overwhelm an
+{es} cluster and bring it down. If you need to connect devices from the internet
+to {es}, use some combination of a proxy, firewall, and VPN. The best
+combination of these technologies depends on your security needs and risks.
+
+[discrete]
+[[security-dont-expose-to-users]]
+=== Implement a barrier between {es} and users
+While you absolutely shouldn't expose {es} directly to the internet, you also
+shouldn't expose {es} directly to users. The following configurations provide
+different methods of user interaction with {es}:
+
+* *Best*: Use an application to make requests on behalf of users.
+* *Better*: If the previous configuration isn't possible, use an application to
+sanitize requests from users.
+* *Good*: If either of the previous configurations isn't possible, implement
```

Comment:
I'll ping @nik9000 again here to help extract the actions users should take for each of these items.

Reply:
At this point I think only the thing you call "best" is acceptable. Everything else is fraught. Back in the day when I wrote that old advice we had folks wanting to just stick a little proxy in front of ES and try to detect bad searches. That is a losing battle.

Whether or not you have a mediating application, tracking who does what is quite useful. Back when I had ES deployed we allowed trusted folks to run some pretty heavy requests. Sometimes we found they hit it too hard and we had to ask them to stop. But we could only do that because we knew them.

Reply:
@nik9000, are there any guidelines that we can provide to users about how to implement an application to make requests on behalf of users? Is there any tangible instruction we can give, or is the an It Depends situation?

Reply:
Yeah, but I don't think I can articulate it well. In principal, the further you get from a proxy the better. If you can, your application should have total control over the kinds of queries and aggs we build. If your users are "the whole internet" then you *have* to have that kind of control. 

For example, you could just plumb it a user's search term to a `match` query or a `match_phrase` query. Or a  `simple_query_string`. I've used `query_string` for this in the past but it was a mistake. Those queries give you less and less control. But `simple_query_string` was designed for this sort of thing so its quite good for it.

Applications that face the whole internet should run pre-canned aggregations. The internet shouldn't really have any influence there. But it's normal to use things like `terms` and `range` for faceting.

If your users are more trusted, say, because they work for you, the you can do the thing kibana does. They let most things through and even help you write complex queries and aggs. This doesn't scale infinitely, but it does scale faster than an organization can hire.

If your users are in between "the internet" and "employees" then you probably should do something in between.

Reply:
> For example, you could just plumb it a user's search term to a `match` query or a `match_phrase` query. Or a `simple_query_string`. I've used `query_string` for this in the past but it was a mistake. Those queries give you less and less control. But `simple_query_string` was designed for this sort of thing so its quite good for it.

Its quite reasonable for applications to parse user queries on their side and then build ES queries from there. Kibana does that. I did a fairly basic version of that in my past. It gives you more control. But it isn't required. Some folks will be fine with `simple_query_string`.